### DISCUSSION TITLE
var和signal之间的差异？

### DISCUSSION THREAD

**Discussion by timimm** [zkp-co-learning/zkp-co-learn#45] [Q&A（提问题在隔壁~）] [answered]
# var和signal之间的差异？
为了搞清楚两者之间的差异，我阅读circom官方文档，希望大佬指正我的理解并解答困惑：
> 菜鸟的理解
> 1. var是便于我们构建约束的变量，它对电路外部是不可见的，也是无法access的（我的理解是恶意的prover无法修改它）
> 2. signal是信号的意思，文档说：“如果在main component中被定义成了public，即`component main {public[a, b]}`，那么它就是对电路外部可见的。对于main component，所有的signal output都必须是public的，否则会报错。”

> 菜鸟的疑问
> 1. `component main = Num2bits(n)`，这样的写法，是意味着Num2bits中所有的signal（包括signal input和signal）都默认public？
> 2. 如果var是不可见的，那么我们在实现isZero电路时是否可以这样写？
```circom
template IsZero () {
    signal input in;
    signal output out;
    
    var inv; // 官方文档是写的signal inv;
    inv <-- in != 0 ? 1/in : 0;
    out <== -inv * in + 1;
    // out*in === 0; 因为 inv 不可修改，无需考虑inv被恶意prover修改的情况。
}
```

**Comment by lkiversonlk** [comment, marked as answer]
var你可以理解成一个占位符，方便你写circom，最后约束系统里如果用到需要展开成原来的样子。
```
    var inv; // 官方文档是写的signal inv;
    inv <-- in != 0 ? 1/in : 0;
    out <== -inv * in + 1;
```

这么写最后你就有一个约束是:

```
out <== (in != 0? 1/in:0) * in + 1;
```

但是这个约束不正确，因为约束力只能有+,*运算

**Reply by timimm** [reply]
谢谢大佬！总结一下`var`和`signal`的差异：
1. signal这种只能赋值一次，而var可以赋值多次（参考Num2bits的accum变量，因为它需要多次赋值，所以是var类型，但它本质上还是signal，所以可以和`in`比较）
2. var类型本质上就是signal量，当它在写入约束系统的时候，需要将其展开成原来的signal量的运算，并检查其是否满足二次算数约束。

**Comment by Mobius3-3** [comment]
补充一下，var的赋值应该用“=”，“<--”只适用于signal。将var用于constraint的时候，要保证var的运算满足quadratic constraints。
